### PR TITLE
feat(integrations): Bitbucket client

### PR DESCRIPTION
**Pull request by lauryndbrown**
Updated the bitbucket client from api 1.0 to 2.0 to be used in Integrations v3 system.

### CODE REVIEW

File: src/sentry/integrations/bitbucket/client.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    repository_diff = u'/2.0/repositories/{username}/{repo_slug}/diff/{spec}'
+    repository_hook = u'/2.0/repositories/{username}/{repo_slug}/hooks/{uid}'
+    repository_hooks = u'/2.0/repositories/{username}/{repo_slug}/hooks'
+
+
+class BitbucketAPI(ApiClient):
+
+    def __init__(self, base_url, shared_secret, subject):
+        # subject is probably the clientKey
+        super(BitbucketAPI, self).__init__(verify_ssl=False)
```
Comment (macqueen) [inline src/sentry/integrations/bitbucket/client.py]: do you need the `verify_ssl` to be false for this to work? because it looks like the old bitbucket client didn't need to

File: src/sentry/integrations/bitbucket/client.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        self.base_url = base_url
+        self.shared_secret = shared_secret
+        self.subject = subject
+
+    # TODO(LB): do has_auth and bind_auth belong here?
```
Comment (macqueen) [inline src/sentry/integrations/bitbucket/client.py]: you probably don't need those methods anymore since the auth method changed

File: src/sentry/integrations/bitbucket/client.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    def request(self, method, path, data=None, params=None, **kwargs):
+        jwt_payload = {
+            'iss': BITBUCKET_KEY,
+            'iat': datetime.datetime.utcnow(),
+            'exp': datetime.datetime.utcnow() + datetime.timedelta(seconds=5 * 60),
+            'qsh': get_query_hash(path, method.upper(), params),
+            'sub': self.subject,
+        }
+        encoded_jwt = jwt.encode(jwt_payload, self.shared_secret)
+        params = dict(**(params or {}))
```
Comment (mnoble) [inline src/sentry/integrations/bitbucket/client.py]: Might be my lack of Python-fu, but is this different than just setting the argument to `data={}` in the method signature?

Comment (lauryndbrown) [reply src/sentry/integrations/bitbucket/client.py]: So, I really hesitate to put dynamic structures in the method signature, due to things being passed by reference, and there are tons of possibilities for bugs. BUT that totally used to do something; there used to be something I was adding to params, and now I'm not. I'll just change this to passing whatever is in params now! Either way good catch!!!

Comment (lauryndbrown) [reply src/sentry/integrations/bitbucket/client.py]: @mnoble 
More to your question pretty sure the pythonic way to handle something like that is:
```
def foo(data=None):
   if data is None:
       data = {}
```
putting dynamic structures in the method signature tends to result in common gotchas! 
http://docs.python-guide.org/en/latest/writing/gotchas/

Comment (mnoble) [reply src/sentry/integrations/bitbucket/client.py]: Makes sense. I don't think it's really a concern when you're using literals as the default value, though.

File: src/sentry/integrations/bitbucket/client.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+
+class BitbucketAPIPath(object):
+    """
+    All UUID's must be surrounded by culybraces.
+    repo_slug - repository slug or UUID
+    username - username or UUID
+    """
+
+    issue = u'/2.0/repositories/{username}/{repo_slug}/issues/{issue_id}'
```
Comment (macqueen) [inline src/sentry/integrations/bitbucket/client.py]: why did you change username and repo slug to be separate? it looks like in the old client they are together https://github.com/getsentry/sentry-plugins/blob/master/src/sentry_plugins/bitbucket/client.py#L34 and a lot of our existing repo / commit functionality depends on their being together

Comment (lauryndbrown) [reply src/sentry/integrations/bitbucket/client.py]: @macqueen that's not the same api. I'm using 2.0 rather than 1.0. If repo_slug actually refers to the username and the slug, it probably should be changed/renamed as that is unclear. I assumed the repo slug only contained the repo slug.

File: src/sentry/integrations/bitbucket/client.py (hunk 1 above)
Comment (lauryndbrown) [reply src/sentry/integrations/bitbucket/client.py]: @macqueen good catch! thank you~!

File: src/sentry/integrations/bitbucket/client.py (hunk 4 above)
Comment (macqueen) [reply src/sentry/integrations/bitbucket/client.py]: ok, feel free to rename it, but the rest of our code combines these so the client is going to have to work with the combined name

Comment (macqueen) [reply src/sentry/integrations/bitbucket/client.py]: screenshot 2018-05-23 14 39 31

Comment (lauryndbrown) [reply src/sentry/integrations/bitbucket/client.py]: So, just so I'm clear the repo_name in this case is username/repo_slug correct? in your example getsentry is the username and sentry is the repo_slug?

Comment (macqueen) [reply src/sentry/integrations/bitbucket/client.py]: yeah, i think it could also probably be `team_slug/repo_slug`? But if you look at the v1 docs here: https://confluence.atlassian.com/bitbucket/issues-resource-296095191.html#issuesResource-GETanindividualissue
we turn this:
`https://api.bitbucket.org/1.0/repositories/{accountname}/{repo_slug}/issues/{issue_id}`
into:
`'/1.0/repositories/%s/issues/%s' % (repo, issue_id)`

so yeah, sorry if that wasn't clear, but by `repo` we mean the whole identifier

**Comment by lauryndbrown** [comment]
@macqueen let me know if the changes are what you expected.

File: src/sentry/integrations/bitbucket/client.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+from sentry.utils.http import absolute_uri
+from sentry.integrations.atlassian_connect import get_query_hash
+from sentry.integrations.client import ApiClient
+
+BITBUCKET_KEY = '%s.bitbucket' % urlparse(absolute_uri()).hostname
+
+
+class BitbucketAPIPath(object):
+    """
+    All UUID's must be surrounded by culybraces.
```
Comment (macqueen) [inline src/sentry/integrations/bitbucket/client.py]: what exactly do you mean by this?

**Review by macqueen — APPROVED** [review]
🎉

File: src/sentry/integrations/bitbucket/client.py (hunk 5 above)
Comment (lauryndbrown) [reply src/sentry/integrations/bitbucket/client.py]: Oh so actually bitbucket (2.0 anyway) will accept UUID's as well. I don't know if outside developers use it and/or if I user really wanted to they could use uuid's as well. https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D 

I can remove that comment if it doesn't make sense for us though!

Comment (lauryndbrown) [reply src/sentry/integrations/bitbucket/client.py]: It's important to note that while a username potentially could change in the future, a `uuid` will not. That's the advantage bitbucket sites of using `uuid`'s.
